File: Source/WebKit2/UIProcess/sling/WebEditableContext.cpp
/*
 * Copyright (C) 2015 Naver Corp. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include "config.h"
#include "WebEditableContext.h"

#include "CompositionUnderline.h"
#include "WebEditable.h"
#include <ManagedVector.h>
#include <util/StringConversion.h>
#include <view/GeometryConversion.h>
#include <view/UIEvent.h>

namespace webkit {

std::shared_ptr<WebEditableContext> WebEditableContext::create(std::shared_ptr<SWKView>& view)
{
    return std::shared_ptr<WebEditableContext>(new WebEditableContext(view));
}

WebEditableContext::WebEditableContext(std::shared_ptr<SWKView>& view)
    : m_editable(WebEditable::create(view))
{
}

WebEditableContext::~WebEditableContext()
{
}

view::Rect WebEditableContext::editableCaretRectAtStart()
{
    return view::geometry_cast<view::Rect>(*m_editable->editableCaretRectAtStart());
}

view::Rect WebEditableContext::editableCaretRectAtEnd()
{
    return view::geometry_cast<view::Rect>(*m_editable->editableCaretRectAtEnd());
}

std::wstring WebEditableContext::editableSelectionText()
{
    return std::s2ws(m_editable->editableSelectionText());
}

std::wstring WebEditableContext::editableTextBeforeCaret(int32_t length)
{
    return std::s2ws(m_editable->editableTextBeforeCaret(length));
}

std::wstring WebEditableContext::editableTextAfterCaret(int32_t length)
{
    return std::s2ws(m_editable->editableTextAfterCaret(length));
}

bool WebEditableContext::setComposition(const std::wstring& text, const std::vector<view::CompositionClause>& clauses, int32_t selectionStart, int32_t selectionEnd, int32_t replacementStart, int32_t replacementEnd)
{
    if (clauses.empty())
        return m_editable->setComposition(std::ws2s(text), nullptr, selectionStart, selectionEnd,
            replacementStart, replacementEnd);

    std::shared_ptr<Managed::Vector> underlines = Managed::Vector::create(clauses.size());
    for (auto& clause : clauses)
        underlines->add(CompositionUnderline::create(clause));
    return m_editable->setComposition(std::ws2s(text), underlines, selectionStart, selectionEnd,
        replacementStart, replacementEnd);
}

bool WebEditableContext::confirmComposition(const std::wstring& text)
{
    return m_editable->confirmComposition(std::ws2s(text));
}

bool WebEditableContext::finishComposition()
{
    return m_editable->finishComposition();
}

void WebEditableContext::cancelComposition()
{
    m_editable->cancelComposition();
}

bool WebEditableContext::performEditorAction(EditorAction editorAction)
{
    return m_editable->performEditorAction(static_cast<int32_t>(editorAction));
}

bool WebEditableContext::sendKeyEvent(const view::UIEvent& event)
{
    if (event.source() != view::UIEvent::Source::Key)
        return false;

    std::string text = std::ws2s(std::wstring(&event.keyState().unicodeCharacter, 1));
    return m_editable->sendKeyEvent(event.timestamp().count(), static_cast<int32_t>(event.action()), event.modifiers().value,
        event.keyState().virtualKeyCode, event.keyState().nativeKeyCode, text, text,
        event.keyState().isAutoRepeat, event.keyState().isSystemKey, event.keyState().isKeypadKey,
        event.keyState().strokeHash);
}

} // namespace webkit

File: Source/WebKit2/UIProcess/sling/WebEditableContext.java
package labs.naver.webkit.view;

import java.util.Vector;

import android.graphics.Rect;
import labs.naver.view.CompositionClause;
import labs.naver.view.EditableContext;
import labs.naver.view.UIEvent;
import labs.naver.webkit.CompositionUnderline;
import labs.naver.webkit.WebEditable;
import labs.naver.webkit.api.SWKView;

public class WebEditableContext implements EditableContext {

    @Override
    public Rect editableCaretRectAtStart() {
        return mEditable.editableCaretRectAtStart();
    }

    @Override
    public Rect editableCaretRectAtEnd() {
        return mEditable.editableCaretRectAtEnd();
    }

    @Override
    public String editableSelectionText() {
        return mEditable.editableSelectionText();
    }

    @Override
    public String editableTextBeforeCaret(int length) {
        return mEditable.editableTextBeforeCaret(length);
    }

    @Override
    public String editableTextAfterCaret(int length) {
        return mEditable.editableTextAfterCaret(length);
    }

    @Override
    public boolean setComposition(String text,
            Vector<CompositionClause> clauses, int selectionStart,
            int selectionEnd, int replacementStart, int replacementEnd) {
        if (clauses == null)
            return mEditable.setComposition(text, null, selectionStart,
                    selectionEnd, replacementStart, replacementEnd);

        Vector<CompositionUnderline> underlines = new Vector<CompositionUnderline>(
                clauses.size());
        for (CompositionClause clause : clauses)
            underlines.add(new CompositionUnderline(clause));
        return mEditable.setComposition(text, underlines, selectionStart,
                selectionEnd, replacementStart, replacementEnd);
    }

    @Override
    public boolean confirmComposition(String text) {
        return mEditable.confirmComposition(text);
    }

    @Override
    public boolean finishComposition() {
        return mEditable.finishComposition();
    }

    @Override
    public void cancelComposition() {
        mEditable.cancelComposition();
    }

    @Override
    public boolean performEditorAction(int editorAction) {
        return mEditable.performEditorAction(editorAction);
    }

    @Override
    public boolean sendKeyEvent(UIEvent event) {
        return mEditable.sendKeyEvent(event.timestamp(), event.action(),
                event.modifiers(), event.keyState().mVirtualKeyCode,
                event.keyState().mNativeKeyCode,
                Character.toString(event.keyState().mUnicodeCharacter),
                Character.toString(event.keyState().mUnicodeCharacter),
                event.keyState().mIsAutoRepeat, event.keyState().mIsSystemKey,
                event.keyState().mIsKeypadKey, event.keyState().mStrokeHash);
    }

    private WebEditable mEditable;

    public WebEditableContext(SWKView view) {
        mEditable = new WebEditable(view);
    }
}
